src/main/java/com/github/manolo8/darkbot/core/utils/factory/EntityListener.java

```diff
@@ -32,49 +37,20 @@ public void setMain(Main main) {
         if (this.main == null) this.main = main;
     }
 
-    public void send(long address, int id) {
-        EntityFactory type = getEntityType(address, id);
-        switch (type) {
-            case PORTAL:
-                int portalType = API.readMemoryInt(address + 112);
-                int x = (int) API.readMemoryDouble(address, 64, 32);
-                int y = (int) API.readMemoryDouble(address, 64, 40);
-
-                send(type, main.starManager.getOrCreate(id, portalType, x, y), address);
-                break;
-            case SHIP:
-                int isNpc = API.readMemoryInt(address + 112);
-
-                if (isNpc == 1) send(EntityFactory.NPC, id, address);
-                else if ((isNpc == 0 && address != main.hero.address &&
-                        address != main.hero.pet.address)) send(type, id, address);
-                break;
-            case BARRIER:
-                String zoneKey = API.readMemoryString(address, 136);
-
-                if (zoneKey.equals("NOA"))      send(type, id, address);
-                else if (zoneKey.equals("DMG")) send(EntityFactory.MIST_ZONE, id, address);
-                break;
-            default:
-                send(type, id, address);
-        }
-    }
 
-    private void send(EntityFactory type, int id, long address) {
-        send(type, type.createEntity(id), address);
-    }
+    public void sendEntity(long address, int id) {
+        EntityFactory type = getEntityType(address, id).get(address);
+        Entity entity;
 
-    private void send(EntityFactory type, Entity entity, long address) {
-        getListener(type).send(onAdd(entity, address));
-    }
+        if (type == EntityFactory.PORTAL) entity = getOrCreatePortal(address, id);
```

Comment:
```suggestion
```
Should be defined as the constrcutor that's called on `createEntity`. Possibly consider making it accept the address not just the id